bridge/tracker/tracker.go

```diff
@@ -0,0 +1,297 @@
+package tracker
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/0xPolygon/polygon-edge/blockchain/storage"
+	"github.com/hashicorp/go-hclog"
+	"github.com/umbracle/go-web3"
+	"math/big"
+)
+
+type Config struct {
+	Confirmations uint64
+	RootchainWS   string
+	DBPath        string
+	ContractABIs  map[string][]string
+}
+
+//	cancellable context for tracker's listening mechanism
+type contextSubscription struct {
+	context context.Context
+	cancel  context.CancelFunc
+}
+
+//	done returns the contexts Done channel.
+func (c *contextSubscription) done() <-chan struct{} {
+	return c.context.Done()
+}
+
+//	Tracker represents an event listener that notifies
+//	for each event emitted by a smart contract on the rootchain.
+//	Events are collected from blocks at the desired depth (confirmations).
+type Tracker struct {
+	logger hclog.Logger
+
+	// required block confirmations
+	confirmations *big.Int
+
+	// event channel
+	eventCh chan []byte
+
+	// cancel subscription
+	ctxSubscription contextSubscription
+
+	//	rootchain subscription object
+	sub subscription
+
+	//	rootchain client
+	client *rootchainClient
+
+	//	db for last processed block number
+	db storage.Storage
+
+	//	events that tracker will listen for
+	contracts []*contractABI
+}
+
+//	NewEventTracker returns a new tracker object.
+func NewEventTracker(logger hclog.Logger, config *Config) (*Tracker, error) {
+	if config == nil {
+		//	load default
+		config = &Config{
+			Confirmations: DefaultBlockConfirmations,
+			RootchainWS:   rootchainWS,
+			DBPath:        "",
+			ContractABIs:  nil,
+		}
+	}
+
+	//	create tracker
+	tracker := &Tracker{
+		logger:        logger.Named("event_tracker"),
+		confirmations: big.NewInt(0).SetUint64(config.Confirmations),
+		eventCh:       make(chan []byte),
+	}
+
+	var err error
+
+	//	load abi events
+	if config.ContractABIs != nil {
+		tracker.contracts = loadABIs(config.ContractABIs)
+	}
+
+	//	load db (last processed block number)
+	if tracker.db, err = initRootchainDB(logger, config.DBPath); err != nil {
+		logger.Error("cannot initialize db", "err", err)
+
+		return nil, err
+	}
+
+	//	create rootchain client
+	if tracker.client, err = newRootchainClient(config.RootchainWS); err != nil {
+		logger.Error("cannot connect to rootchain", "err", err)
+
+		return nil, err
+	}
+
+	//	subscribe for new headers
+	if tracker.sub, err = tracker.client.subscribeNewHeads(); err != nil {
+		logger.Error("cannot subscribe to rootchain", "err", err)
+	}
```

Comment:
Nitpick. I think better to start to subscribe after Start() is called, not in constructor

Reply:
[Moved](https://github.com/0xPolygon/polygon-edge/pull/386/commits/d8ba42ed44ff9eb623c01b2ff90942b5530d7ebb)